### ISSUE TITLE
Rolling restart of a nomad job

### ISSUE THREAD

**Issue by shishir-a412ed** [hashicorp/nomad#12490]
Hi!

We are trying to formulate a process internally on what would be the best way to do rolling restarts on a nomad job. Currently we give our users two options 

(1) To change a `meta` variable in the Nomad job spec and trigger a new deployment. This results in a pseudo deployment which causes Nomad to restarts all the allocations. Users prefer this for two reasons:
       - There is a rollback mechanism to the previous deployment if something goes wrong.
       - We use Consul with Nomad for service discovery. When they rollout a new deployment, the service is first de-registered from Consul ---> wait for [`shutdown_delay`](https://www.nomadproject.io/docs/job-specification/task#shutdown_delay) ---> restart the allocation (SIGTERM + graceful shutdown + SIGKILL). This is really important for them because they don't want to keep receiving traffic while the restart is in place, and consul service de-registration helps them with that usecase.

(2) We have an internal version of `nomad-job-restart` which allows them to restart all the allocations in a job in batches with fixed wait.
Let's say we have a job `foobar` with 100 allocations.

e.g. `nomad job restart --batch-size 10 --batch-wait 5 foobar` will restart 10 allocations at a time and wait for 5 seconds between those batch restarts.

While the long term goal is to get https://github.com/hashicorp/nomad/pull/11533 in. @hollowsunsets is helping with some of that work.

Short term, I feel they should not be using re-deployments to trigger restarts at massive scale (Option 1) e.g. doing re-deployment to do restarts for a few jobs is fine. However during an outage if we need to do rolling restart on all our services, I would rather have them use (2) [in-place-restart] than (1) since I feel that puts some pressure on the scheduler since now the scheduler would have to run through node matching and scoring algorithms and find new placements for these new deployments whose actual goal is just to restart their services. In usecases like these I feel an in-place restart is much lighter.

Here are my questions:

- Is my concern valid? Should we encourage the user to use in-place restarts (1) over re-deployments (2) when their goal is to just restart their services?
- Right now, in place restarts (nomad alloc restart) doesn't respect `shutdown_delay`. Will you be open to a PR where if a user does an in-place restart of an allocation, the corresponding service gets de-registered from consul first, so the container stop receiving traffic while the restart is in place. I guess this makes more sense at the job level than at the allocation level, but would be good to get your thoughts on this.

@schmichael Tagging you since you have the most background on this 🙂

**Comment by tgross** [comment]
> Is my concern valid? Should we encourage the user to use in-place restarts (1) over re-deployments (2) when their goal is to just restart their services?

You're definitely right here. At large scale, restarting is _much_ cheaper that rescheduling. Restarting happens almost entirely client-side. But in addition to avoiding the scheduler load, you also avoid having to pull new workload images to any hosts where they don't exist, you don't have to mint new Vault tokens, etc. so that's less work for all the allocation's dependencies as well.

> Right now, in place restarts (nomad alloc restart) doesn't respect `shutdown_delay`. Will you be open to a PR where if a user does an in-place restart of an allocation, the corresponding service gets de-registered from consul first, so the container stop receiving traffic while the restart is in place. I guess this makes more sense at the job level than at the allocation level, but would be good to get your thoughts on this.

Interesting... the `nomad alloc restart` kills all the tasks in parallel, so I suspect the group `shutdown_delay` is only getting called when the whole alloc is going down. But tasks can have their own [`shutdown_delay`](https://www.nomadproject.io/docs/job-specification/task#shutdown_delay) too. Does this not work as you'd expect already?

**Comment by shishir-a412ed** [comment]
@tgross Apologies for the delayed response. I was able to verify this behaviour with in-place restarts (`nomad alloc restart`) and it looks like it works exactly the same as the deployment workflow (which people in our organization are using to trigger a restart).

I deployed the following job

```
job "signal-job" {
  datacenters = ["dc1"]

  group "sgroup" {
    count=3
    task "stask" {
      driver = "docker"

      config {
        image = "shm32/signal_handler:1.0"
      }

      resources {
        cpu    = 600
        memory = 512
      }
    }

    shutdown_delay = "120s"

    service {
       name = "${JOB}"
    }
  }
}
```

- This job started 3 allocations (3 containers) and registered 3 corresponding services in consul.
- I did a `nomad alloc restart` on 1 of them.
- Consul immediately de-registered the service associated with that allocation (So that no more production traffic is routed to that container while the restart is in-place)
- Nomad then waited for `120s` to honor the [`shutdown_delay`](https://www.nomadproject.io/docs/job-specification/task#shutdown_delay)
- Nomad send a `SIGTERM` to the container
- Wait for 5 seconds (default [`kill_timeout`](https://www.nomadproject.io/docs/job-specification/task#kill_timeout))
- Send `SIGKILL` and eventually restart the container.
- Register the service back with Consul.

I am going to close this issue as in-place restarts works exactly like pseudo deployments (creating a new deployment by changing a meta variable) to trigger restarts, however in-place restarts should be preferred over deployment workflow if one just wants to do a rolling restart, as they are more lightweight and puts less pressure on the Nomad server.
